File: srcs/requirements/Pong-Server/server/game/Ball.py
from config import *
from Vec2 import *

class Ball:
	def __init__(self, borderless):
		self.ai_hitbox = False
		self.borderless = borderless
		self.radius = 7
		self.center = [Vec2(winWidth / 2, winHeight / 2)]
		self.speed = ball_speed_per_sec * 0.005
		if borderless:
			self.center.append(Vec2(self.center[0].x, self.center[0].y + winHeight))
			self.center.append(Vec2(self.center[0].x, self.center[0].y - winHeight))

		self.stick = 0
		self.side = "none"
		self.launch()
  
		self.multiplier = 1.0
		self.last_hit = 0


	def copy(self, ai_hitbox = False):
		ball = Ball(self.borderless)
		ball.center = [Vec2(pos=self.center[0])]
		if self.borderless:
			ball.center.append(Vec2(pos=self.center[1]))
			ball.center.append(Vec2(pos=self.center[2]))
		ball.stick = self.stick
		ball.side = self.side
		ball.dir = self.dir
		ball.multiplier = self.multiplier
		ball.last_hit = self.last_hit
		ball.ai_hitbox = ai_hitbox
		return ball


	def move(self, players, walls, obstacle):
		if self.stick != 0:
			for player in players:
				if player.nb == self.stick:
					self.side = player.side
					if player.side == "left":
						self.center[0] = Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) + 25, player.paddle[0].pos.y + (player.size[1] / 2))
					if player.side == "right":
						self.center[0] = Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) - 25, player.paddle[0].pos.y + (player.size[1] / 2))
					if player.side == "up":
						self.center[0] = Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) + 25)
					if player.side == "down":
						self.center[0] = Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) - 25)
			return

		rad = radians(self.dir)
		if self.radius > self.speed:
			new_x = self.center[0].x + (self.speed * cos(rad))
			new_y = self.center[0].y + (self.speed * sin(rad))
			self.center[0].x = new_x
			self.center[0].y = new_y
			return

		tmp_speed = self.radius
		collision = False
		while not collision and tmp_speed <= self.speed:
			for center in self.center:
				tmp_x = center.x + (tmp_speed * cos(rad))
				tmp_y = center.y + (tmp_speed * sin(rad))
				collision = try_collide(tmp_x, tmp_y, self.radius, players, walls, obstacle, self.ai_hitbox)
				if collision:
					break
			tmp_speed += self.radius

		if collision:
			new_x = tmp_x
			new_y = tmp_y
		else:
			new_x = self.center[0].x + (self.speed * cos(rad))
			new_y = self.center[0].y + (self.speed * sin(rad))

		self.center[0].x = new_x
		self.center[0].y = new_y
  
	def collide(self, walls, players, obstacle):
		if obstacle:
			if getDist(self.center[0], obstacle.center) <= self.radius + obstacle.radius and obstacle.solid:
				obstacle.collide(self)
		if walls:
			for wall in walls:
				if not wall.square:
					wall.collide(self)
		if not self.ai_hitbox:
			for player in players:
				player.collide(self)

  
	def update(self, core, delta):
		self.speed = ball_speed_per_sec * delta * self.multiplier
		self.move(core.players, core.walls, core.obstacle)
		self.collide(core.walls, core.players, core.obstacle)
		if not self.ai_hitbox:
			self.goal(core.players, core.custom_mod)
		self.unstuck(core.custom_mod)

		if self.borderless:
			if self.center[0].y < 0:
				self.center[0].y += winHeight
			if self.center[0].y >= winHeight:
				self.center[0].y -= winHeight
			self.center[1].x = self.center[0].x
			self.center[1].y = self.center[0].y - winHeight
			self.center[2].x = self.center[0].x
			self.center[2].y = self.center[0].y + winHeight

	def unstuck(self, mod):
		if mod == "1V1V1V1":
			return
		if round(self.dir) % 360 in range(85, 96) or round(self.dir) % 360 in range(-275, -264):
			if (self.last_hit == 1):
				self.dir -= 5
			else:
				self.dir += 5
		if round(self.dir) % 360 in range(265, 276) or round(self.dir) % 360 in range(-95, -86):
			if (self.last_hit == 1):
				self.dir += 5
			else:
				self.dir -= 5
	
	def goal(self, players, mod):
		ball_box = [self.center[0].x - self.radius, self.center[0].y - self.radius]
		for player in players:
			if is_colliding(ball_box, [self.radius * 2, self.radius * 2], player.goal.pos, player.goal.size):
				if players.__len__() == 2:
					if player.nb == 1:
						players[1].score += 1
					else:
						players[0].score += 1
				elif mod == "1V1V1V1":
					if player.nb == self.last_hit:
						player.score -= 1
					elif self.last_hit:
						players[self.last_hit - 1].score += 1
					else:
						out = player.nb
						for other in players:
							if other.nb != out:
								other.score += 1
				elif players.__len__() == 4:
					if player.nb == 1 or player.nb == 2:
						players[2].score += 1
					else:
						players[0].score += 1
				self.stick = player.nb
				self.multiplier = 1.0
				self.side = player.side
				if player.side == "left":
					self.center[0] = Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) + 25, player.paddle[0].pos.y + (player.size[1] / 2))
				if player.side == "right":
					self.center[0] = Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) - 25, player.paddle[0].pos.y + (player.size[1] / 2))
				if player.side == "up":
					self.center[0] = Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) + 25)
				if player.side == "down":
					self.center[0] = Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) - 25)
				break

	def launch(self):
		if self.stick == 0:
			self.dir = randint(0, 359)
			while self.dir in range(75, 106) or self.dir in range(255, 286):
				self.dir = randint(0, 359)
		else:
			if self.side == "left":
				self.dir = 0
			elif self.side == "right":
				self.dir = 180
			elif self.side == "up":
				self.dir = 90
			elif self.side == "down":
				self.dir = 270
	  
		self.multiplier = 1.0
		self.stick = 0
  
  
def try_collide(x, y, radius, players, walls, obstacle, ai_hitbox = False):
	tmp = [x - radius, y - radius]
	
	for player in players:
		for paddle in player.paddle:
			if is_colliding(tmp, [radius * 2, radius * 2], paddle.pos, player.size):
				return True

	if obstacle and getDist(Vec2(x, y), obstacle.center) <= radius + obstacle.radius and obstacle.solid:
		return True

	if ai_hitbox:
		if is_colliding(tmp, [radius * 2, radius * 2], ai_hitbox.pos, ai_hitbox.size):
			return True

	if not walls:
		return False
	for wall in walls:
		if not wall.square and is_colliding(tmp, [radius * 2, radius * 2], wall.hitbox.pos, wall.size):
			return True
 
	return False

File: srcs/django_code/front/js/game/Ball.js
import { Vec2, getDist } from "./Vec2.js";
import { canvas, ctx, ctx_circle } from "./canvas.js";
import { is_colliding } from "./Hitbox.js";

export class Ball {
	constructor(borderless) {
		this.speed_per_sec = canvas.width / 3;
		this.speed = this.speed_per_sec * 0.01;
		this.ai_hitbox = false;
		this.borderless = borderless;
		this.radius = Math.floor(canvas.height * 0.01);
		this.center = [new Vec2(canvas.width / 2, canvas.height / 2)];
		if (borderless) {
			this.center.push(new Vec2(this.center[0].x, this.center[0].y + canvas.height));
			this.center.push(new Vec2(this.center[0].x, this.center[0].y - canvas.height));
		}

		this.stick = 0;
		this.side = "none";
		this.launch();
	}

	copy(ai_hitbox = false) {
		let ball = new Ball(this.borderless);
		ball.center = [new Vec2(this.center[0].x, this.center[0].y)];
		if (this.borderless) {
			ball.center.push(new Vec2(this.center[1].x, this.center[1].y));
			ball.center.push(new Vec2(this.center[2].x, this.center[2].y));
		}
		ball.stick = this.stick;
		ball.side = this.side;
		ball.dir = this.dir;
		ball.multiplier = this.multiplier;
		ball.last_hit = this.last_hit;
		ball.ai_hitbox = ai_hitbox;
		return ball;
	}

	move(players, walls, obstacle) {
		if (this.stick != 0) {
			for (let player of players) {
				if (player.nb == this.stick) {
					this.side = player.side;
					if (player.side === "left")
						this.center[0] = new Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) + (this.radius * 4), player.paddle[0].pos.y + (player.size[1] / 2));
					else if (player.side === "right")
						this.center[0] = new Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) - (this.radius * 4), player.paddle[0].pos.y + (player.size[1] / 2));
					else if (player.side === "up")
						this.center[0] = new Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) + (this.radius * 4));
					else
						this.center[0] = new Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) - (this.radius * 4));
				}
			}
			return;
		}

		let rad = this.dir * Math.PI / 180;
		let new_x, new_y;
		if (this.radius > this.speed) {
			new_x = this.center[0].x + (Math.cos(rad) * this.speed);
			new_y = this.center[0].y + (Math.sin(rad) * this.speed);
			this.center[0] = new Vec2(new_x, new_y);
			return;
		}

		let tmp_speed = this.radius;
		let collision = false;
		while (!collision && tmp_speed <= this.speed) {
			for (let center of this.center) {
				new_x = center.x + (Math.cos(rad) * tmp_speed);
				new_y = center.y + (Math.sin(rad) * tmp_speed);
				collision = try_collide(new_x, new_y, this.radius, players, walls, obstacle, this.ai_hitbox);
				if (collision)
					break;
			}
			tmp_speed += this.radius;
		}
		if (!collision) {
			new_x = this.center[0].x + (Math.cos(rad) * this.speed);
			new_y = this.center[0].y + (Math.sin(rad) * this.speed);
		}
		this.center[0] = new Vec2(new_x, new_y);
	}

	collide(walls, players, obstacle) {
		if (obstacle)
			if (getDist(this.center[0], obstacle.center) <= this.radius + obstacle.radius && obstacle.solid)
				obstacle.collide(this);
		if (walls)
			for (let wall of walls)
				if (!wall.square)
					wall.collide(this);
		if (!this.ai_hitbox)
			for (let player of players)
				player.collide(this);
	}

	update(core, delta) {
		this.speed = this.speed_per_sec * delta * this.multiplier;
		this.move(core.players, core.walls, core.obstacle);
		this.collide(core.walls, core.players, core.obstacle);
		if (!this.ai_hitbox)
			this.goal(core.players, core.square);
		this.unstuck(core.square);

		if (this.borderless) {
			if (this.center[0].y < 0)
				this.center[0].y += canvas.height;
			else if (this.center[0].y > canvas.height)
				this.center[0].y -= canvas.height;
			this.center[1].x = this.center[0].x;
			this.center[1].y = this.center[0].y + canvas.height;
			this.center[2].x = this.center[0].x;
			this.center[2].y = this.center[0].y - canvas.height;
		}
	}

	draw() {
		for (let center of this.center)
			ctx_circle(ctx, center.x, center.y, this.radius).fill();
	}

	unstuck(mod) {
		if (mod)
			return;
		if ((Math.round(this.dir) % 360 >= 85 && Math.round(this.dir) % 360 <= 95) || (Math.round(this.dir) % 360 >= -275 && Math.round(this.dir) % 360 <= -265)) {
			if (this.last_hit == 1)
				this.dir -= 5;
			else
				this.dir += 5;
		}
		if ((Math.round(this.dir) % 360 >= 265 && Math.round(this.dir) % 360 <= 275) || (Math.round(this.dir) % 360 >= -95 && Math.round(this.dir) % 360 <= -85)) {
			if (this.last_hit == 1)
				this.dir += 5;
			else
				this.dir -= 5;
		}

	}

	goal(players, mod) {
		let ball_box = new Vec2(this.center[0].x - this.radius, this.center[0].y - this.radius);
		for (let player of players) {
			if (is_colliding(ball_box, [this.radius * 2, this.radius * 2], player.goal.pos, player.goal.size)) {
				this.draw();
				if (players.length === 2) {
					if (player.nb === 1)
						players[1].score += 1;
					else
						players[0].score += 1;
				}
				else if (mod) {
					if (player.nb == this.last_hit)
						player.score -= 1;
					else if (this.last_hit)
						players[this.last_hit - 1].score += 1;
					else
						for (let other of players)
							if (other.nb != player.nb) {
								other.score += 1;
							}
				}
				else if (players.length === 4) {
					if (player.nb === 1 || player.nb === 2) {
						players[2].score += 1;
						players[3].score += 1;
					}
					else {
						players[0].score += 1;
						players[1].score += 1;
					}
				}
				this.stick = player.nb;
				this.multiplier = 1.0;
				this.side = player.side;
				if (player.side === "left")
					this.center[0] = new Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) + (this.radius * 2), player.paddle[0].pos.y + (player.size[1] / 2));
				else if (player.side === "right")
					this.center[0] = new Vec2((player.paddle[0].pos.x + (player.size[0] / 2)) - (this.radius * 2), player.paddle[0].pos.y + (player.size[1] / 2));
				else if (player.side === "up")
					this.center[0] = new Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) + (this.radius * 2));
				else
					this.center[0] = new Vec2(player.paddle[0].pos.x + (player.size[0] / 2), (player.paddle[0].pos.y + (player.size[1] / 2)) - (this.radius * 2));
				break;
			}
		}
	}

	launch() {
		if (this.stick == 0) {
			this.dir = randInt(0, 360);
			while ((this.dir >= 75 && this.dir <= 105) || (this.dir >= 255 && this.dir <= 285))
				this.dir = randInt(0, 360);
		}
		else {
			if (this.side === "left")
				this.dir = 0;
			else if (this.side === "right")
				this.dir = 180;
			else if (this.side === "up")
				this.dir = 90;
			else
				this.dir = 270;
		}
		this.last_hit = this.stick;
		this.stick = 0;
		this.multiplier = 1.0;
	}

	responsive(old_sizes) {
		this.speed_per_sec = canvas.width / 3;
		this.radius = Math.floor(canvas.height * 0.01);
		const pos_ratio = [this.center[0].x / old_sizes[0], this.center[0].y / old_sizes[1]];
		this.center[0] = new Vec2(pos_ratio[0] * canvas.width, pos_ratio[1] * canvas.height);
		if (this.borderless) {
			if (this.center[0].y < 0)
				this.center[0].y += canvas.height;
			else if (this.center[0].y > canvas.height)
				this.center[0].y -= canvas.height;
			this.center[1].x = this.center[0].x;
			this.center[1].y = this.center[0].y + canvas.height;
			this.center[2].x = this.center[0].x;
			this.center[2].y = this.center[0].y - canvas.height;
		}
	}
}

function randInt(min, max) {
	return Math.floor(Math.random() * (max - min) + min);
}

export function try_collide(x, y, radius, players, walls, obstacle, ai_hitbox = false) {
	let ball_box = new Vec2(x - radius, y - radius);

	for (let player of players)
		for (let paddle of player.paddle)
			if (is_colliding(ball_box, [radius * 2, radius * 2], paddle.pos, player.size))
				return true;
	
	if (obstacle && getDist(new Vec2(x, y), obstacle.center) <= radius + obstacle.radius && obstacle.solid)
		return true;

	if (ai_hitbox)
		if (is_colliding(ball_box, [radius * 2, radius * 2], ai_hitbox.pos, ai_hitbox.size))
			return true;

	if (!walls)
		return false;
	for (let wall of walls)
		if (!wall.square && is_colliding(ball_box, [radius * 2, radius * 2], wall.Hitbox.pos, wall.size))
			return true;
	return false;
}